### PR TITLE
Alternatives to recall

### PR DESCRIPTION
**Pull request by bills7281**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/domain/makerecalldecisions/recommendation/AlternativesToRecallTried.kt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package uk.gov.justice.digital.hmpps.makerecalldecisionapi.domain.makerecalldecisions.recommendation
+
+data class AlternativesToRecallTried(
+  val selected: List<SelectedAlternative>?,
+  val allOptions: List<RecallAlternative>?
```
Comment (jackmaloney) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/domain/makerecalldecisions/recommendation/AlternativesToRecallTried.kt:5]: You can use `List<TextValueOption>?` here and remove your `RecallAlternative` class

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/util/RecommendationToPartADataMapper.kt [hunk 2]
```diff
@@ -16,7 +17,25 @@ class RecommendationToPartADataMapper {
...
-      return PartAData(custodyStatus, recallType, recommendation.data.responseToProbation, isThisAnEmergencyRecall)
+      val selectedAlternativesMap = convertToMap(recommendation.data.alternativesToRecallTried?.selected)
+
+      return PartAData(custodyStatus, recallType, recommendation.data.responseToProbation, isThisAnEmergencyRecall, selectedAlternativesMap)
+    }
+
+    private fun convertToMap(selectedAlternatives: List<SelectedAlternative>?): Map<String, String> {
+      val selectedAlternativesMap = selectedAlternatives?.associate { it.value to it.details } ?: emptyMap()
+      return mapOf(
+        "warning_letter_details" to (selectedAlternativesMap["WARNINGS_LETTER"] ?: ""),
```
Comment (jackmaloney) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/util/RecommendationToPartADataMapper.kt:28]: You can use https://github.com/ministryofjustice/make-recall-decision-api/blob/main/src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/util/MrdTextConstants.kt#L6 here maybe

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/util/RecommendationToPartADataMapper.kt [hunk 3]
```diff
@@ -16,7 +17,25 @@ class RecommendationToPartADataMapper {
       val recallType = findRecallTypeToDisplay(recommendation.data.recallType)
       val isThisAnEmergencyRecall = convertBooleanToYesNo(recommendation.data.isThisAnEmergencyRecall)
 
-      return PartAData(custodyStatus, recallType, recommendation.data.responseToProbation, isThisAnEmergencyRecall)
+      val selectedAlternativesMap = convertToMap(recommendation.data.alternativesToRecallTried?.selected)
+
+      return PartAData(custodyStatus, recallType, recommendation.data.responseToProbation, isThisAnEmergencyRecall, selectedAlternativesMap)
+    }
+
+    private fun convertToMap(selectedAlternatives: List<SelectedAlternative>?): Map<String, String> {
```
Comment (jackmaloney) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/util/RecommendationToPartADataMapper.kt:25]: Is it worth changing the name of this function to be specific for the SelectedAlternative?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/integration/controller/RecommendationControllerTest.kt [hunk 4]
```diff
@@ -136,6 +137,52 @@ class RecommendationControllerTest() : IntegrationTestBase() {
       .isEqualTo("No recommendation available: No recommendation found for id: 999")
   }
 
+  @Test
+  fun `update a recommendation with alternatives to recall`() {
+    userAccessAllowed(crn)
+    allOffenderDetailsResponse(crn)
+    deleteAndCreateRecommendation()
+    updateRecommendation(updateRecommendationRequest())
+    updateRecommendation(updateRecommendationAlternativesToRecallRequest())
```
Comment (jackmaloney) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/integration/controller/RecommendationControllerTest.kt:146]: Not sure you need this? Can we not just do it through the current pattern of putting the update into 'updateRecommendationRequest()' and then triggering 'secondUpdateRecommendationRequest()'? Otherwise we're going to end up with lots of duplicated test data.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/PartATemplateReplacementServiceTest.kt [hunk 5]
```diff
@@ -46,16 +53,36 @@ internal class PartATemplateReplacementServiceTest : ServiceTestBase() {
...
       assertThat(result["is_this_an_emergency_recall"]).isEqualTo("Yes")
+      assertThat(result["warning_letter_details"]).isEqualTo("We sent a warning letter on 27th July 2022")
+      assertThat(result["drug_testing_details"]).isEqualTo("drugs test passed")
+      assertThat(result["increased_frequency_details"]).isEqualTo("increased frequency")
+      assertThat(result["extra_licence_conditions_details"]).isEqualTo("licence conditions added")
+      assertThat(result["referral_to_other_teams_details"]).isEqualTo("referral to other team")
+      assertThat(result["referral_to_approved_premises_details"]).isEqualTo("referred to approved premises")
+      assertThat(result["referral_to_partnership_agencies_details"]).isEqualTo("referred to partner agency")
+      assertThat(result["risk_escalation_details"]).isEqualTo("risk escalation")
+      assertThat(result["alternative_to_recall_other_details"]).isEqualTo("alternative action")
```
Comment (jackmaloney) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/PartATemplateReplacementServiceTest.kt:85]: Is it possible to add a test for the 'None' with details route to this test class?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/domain/makerecalldecisions/recommendation/AlternativesToRecallTried.kt [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+)
+
+data class SelectedAlternative(
+  val value: String?,
+  val details: String?
+)
+
+data class RecallAlternative(
```
Comment (jonnywyatt) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/domain/makerecalldecisions/recommendation/AlternativesToRecallTried.kt:13]: I think @jackmaloney created a shared type for this already

File: src/test/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/RecommendationServiceTest.kt [hunk 7]
```diff
@@ -110,7 +117,8 @@ internal class RecommendationServiceTest : ServiceTestBase() {
           lastModifiedDate = "2022-07-26T09:48:27.443Z",
           lastModifiedBy = "Bill",
           createdBy = existingRecommendation.data.createdBy,
-          createdDate = existingRecommendation.data.createdDate
+          createdDate = existingRecommendation.data.createdDate,
+          alternativesToRecallTried = existingRecommendation.data.alternativesToRecallTried
```
Comment (jackmaloney) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/RecommendationServiceTest.kt:121]: Is it possible to add a test for the 'None' with details route to this test class? It doesn't end up in the Part A but we need to make sure it's saved in the database and retrieved in save and return route and I can't see any tests covering this.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/integration/requests/makerecalldecisions/UpdateRecommendationAlternativesToRecallRequest.kt [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+fun updateRecommendationAlternativesToRecallRequest() = """
+{
+	"alternativesToRecallTried": {
+		"selected": [{
+				"value": "WARNINGS_LETTER",
+				"details": "We sent a warning letter on 27th July 2022"
+			},
+			{
+				"value": "DRUG_TESTING"
```
Comment (jonnywyatt) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/integration/requests/makerecalldecisions/UpdateRecommendationAlternativesToRecallRequest.kt:11]: I think details is mandatory for all values?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/domain/makerecalldecisions/recommendation/AlternativesToRecallTried.kt (hunk 1 above)
Comment (bills7281) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/domain/makerecalldecisions/recommendation/AlternativesToRecallTried.kt]: Done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/util/RecommendationToPartADataMapper.kt (hunk 2 above)
Comment (bills7281) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/util/RecommendationToPartADataMapper.kt]: Done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/util/RecommendationToPartADataMapper.kt (hunk 3 above)
Comment (bills7281) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/util/RecommendationToPartADataMapper.kt]: Done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/integration/controller/RecommendationControllerTest.kt (hunk 4 above)
Comment (bills7281) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/integration/controller/RecommendationControllerTest.kt]: Neat point, done!

File: src/test/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/RecommendationServiceTest.kt (hunk 7 above)
Comment (bills7281) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/RecommendationServiceTest.kt:125]: Done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/domain/makerecalldecisions/recommendation/AlternativesToRecallTried.kt (hunk 6 above)
Comment (bills7281) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/domain/makerecalldecisions/recommendation/AlternativesToRecallTried.kt]: Done ...twice!!

File: src/test/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/integration/requests/makerecalldecisions/UpdateRecommendationAlternativesToRecallRequest.kt (hunk 8 above)
Comment (bills7281) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/integration/requests/makerecalldecisions/UpdateRecommendationAlternativesToRecallRequest.kt]: Added!

**Review by jonnywyatt — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/util/RecommendationToPartADataMapper.kt [hunk 9]
```diff
@@ -20,7 +21,24 @@ class RecommendationToPartADataMapper {
...
-        dateVloInformed = convertLocalDateToReadableDate(recommendation.data.dateVloInformed)
+        dateVloInformed = convertLocalDateToReadableDate(recommendation.data.dateVloInformed),
+        selectedAlternativesMap = convertToSelectedAlternativesMap(recommendation.data.alternativesToRecallTried?.selected)
+      )
+    }
+
+    private fun convertToSelectedAlternativesMap(selectedAlternatives: List<SelectedAlternative>?): Map<String, String> {
+      val selectedAlternativesMap = selectedAlternatives?.associate { it.value to it.details } ?: emptyMap()
+      return mapOf(
+        "warning_letter_details" to (selectedAlternativesMap["WARNINGS_LETTER"] ?: EMPTY_STRING),
```
[the remark this answers is not in the document]
Comment (jackmaloney) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/util/RecommendationToPartADataMapper.kt:32]: Also, are these mappings stored anywhere? Can't see anything on the ticket so might be worth adding?

**Review by jackmaloney — CHANGES_REQUESTED** [review]
Don't think the out_template.docx needs to be committed. [out_template.docx](https://github.com/ministryofjustice/make-recall-decision-api/pull/296/files#diff-ba3bbce0bdc31ce2c2358864fe89ac02581ca9803bbba5c5c0f65a81fc86357b)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/util/RecommendationToPartADataMapper.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (bills7281) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/util/RecommendationToPartADataMapper.kt]: Done!
I have added a note to the BE ticket and tagged Jon - https://dsdmoj.atlassian.net/browse/MRD-471

**Review by jackmaloney — APPROVED** [review]
